# HANDS-ON EXERCISE 1
## Business Analytics — Sales Performance Calculator

### Scenario

A growing retail business recorded its daily sales for one week.

The goal is to use NumPy to analyze the business's weekly sales performance.

In [5]:
import numpy as np

## Task 1 - Create the dataset

In [6]:
daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])

print(f"""
Daily Sales = {daily_sales}
Type = {type(daily_sales)}
Shape = {daily_sales.shape}
""")


Daily Sales = [125000 150000 175000 140000 190000 210000 160000]
Type = <class 'numpy.ndarray'>
Shape = (7,)



## Explanation

The dataset contains 7 values because the business recorded sales for 7 days.

The data is stored as a NumPy array. Its shape is (7,), meaning it is a one-dimensional array containing 7 elements.

## Task 2 - Calculate Total Sales

In [7]:
total_sales = np.sum(daily_sales)

print(f"Total Sales: {total_sales}")

Total Sales: 1150000


## Explanation



## Task 3 - Calculate Average Daily Sales

In [9]:
average_sales = np.mean(daily_sales)

print(f"Average Daily Sales = {average_sales:.2f}")

Average Daily Sales = 164285.71


## Task 4 - Apply a Sales Adjustment

In [10]:
adjusted_sales = daily_sales * 1.10

print(f"Adjusted Sales: {adjusted_sales}")

Adjusted Sales: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]


## Explanation

The entire array was multiplied by 1.10.

This represents the original sales plus a 10% increase for each day.

## Task 5 - Compare Original and Adjusted Sales

In [13]:
additional_revenue = adjusted_sales - daily_sales

print(f"Additional Sales: {additional_revenue}")

Additional Sales: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]


## Explanation

The additional revenue represents the amount each day's sales would increase by under the hypothetical 10% increase.

For example, the first day's original sales were ₦125,000.

A 10% increase gives:

₦125,000 × 10% = ₦12,500

Therefore, the additional revenue for that day is ₦12,500.

## Task 6 - Use NumPy's Mathematical Functions

In [15]:
average_sales_check = np.mean(daily_sales)

print(f"Total Sales: {total_sales}")
print(f"Average Daily Sales: {average_sales_check:.2f}")

Total Sales: 1150000
Average Daily Sales: 164285.71


## Explanation


The total sales figure tells the business how much revenue was generated across the entire week. The average daily sales figure tells the business how much revenue was generated on a typical day.

The total is useful for understanding overall weekly performance, while the average makes it easier to understand daily performance and compare this week with other periods.

## Business Thinking Challenge


The business generated total weekly sales of ₦1,150,000, with an average daily sales value of approximately ₦164,285.71. This shows the overall revenue generated during the week as well as the typical amount generated per day. The business can use these figures to monitor sales performance and compare future weeks.

# HANDS-ON EXERCISE 2
## Education Analytics — Student Performance Analysis

### Scenario

A training program recorded the scores of a group of students.

The goal is to analyze the students' overall performance and determine how individual scores differ from the group average.

## Task 1 - Create and Inspect the Data

In [16]:
import numpy as np

In [19]:
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])

print(f"""
Scores: {scores}
Type: {type(scores)}
Shape: {scores.shape}
""")


Scores: [62 75 81 69 88 94 73 85 77 91]
Type: <class 'numpy.ndarray'>
Shape: (10,)



## Task 2 - Calculate the Mean Score

In [20]:
mean_score = np.mean(scores)

print(f"Mean Score: {mean_score}")

Mean Score: 79.5


## Task 3 - Calculate Each Student's Deviation

In [22]:
deviations = scores - mean_score

print(f"Deviations: {deviations}")

Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


## Explanation

Each deviation shows how far a student's score is from the group average of 79.5.

A negative value means the score is below the average, while a positive value means the score is above the average.

## Task 4 - Identify Positive and Negative Deviations

## Explanation

A positive deviation means that the student's score is above the group average.

A negative deviation means that the student's score is below the group average.

A value close to zero means that the student's score is close to the group average.

For example, the score of 94 has a deviation of +14.5, meaning it is 14.5 points above the average. The score of 62 has a deviation of -17.5, meaning it is 17.5 points below the average.

## Task 5 - Square the Deviations

In [23]:
squared_deviations = deviations ** 2

print(f"Squared Deviations: {squared_deviations}")

Squared Deviations: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]


## Explanation


Squaring a number makes a negative value positive because a negative number multiplied by another negative number produces a positive result.

For example:

(-17.5)² = 306.25

This allows the deviations to be combined without positive and negative values cancelling each other out.

## Task 6 - Calculate the sum of squared deviations

In [24]:
sum_squared_deviations = np.sum(squared_deviations)

print(f"Sum of the Squared Deviations: {sum_squared_deviations}")

Sum of the Squared Deviations: 932.5


## Task 7 - Connect the Steps to Standard Deviation

## Connecting the Steps to Standard Deviation

The process of calculating standard deviation can be understood through these stages:

1. Calculate the mean.
2. Calculate the deviations of each value from the mean.
3. Square the deviations.
4. Sum the squared deviations.
5. Continue with the appropriate division.
6. Take the square root.

In this exercise:

- Mean = 79.5
- Sum of squared deviations = 932.5

Standard deviation is useful because it helps show how spread out the student scores are around the average. While the mean tells us the typical score, standard deviation gives additional information about the variation in the scores.

## Industry Thinking Challenge


The average score gives a general idea of the group's performance, but it does not show how individual students performed relative to that average. Looking at deviations helps identify students whose scores are considerably above or below the average. This can help the program coordinator understand differences in student performance and identify areas that may require additional attention.

# HANDS-ON EXERCISE 3
## Engineering & Scientific Computing — Trigonometric Series and Convergence

### Scenario

An engineering/scientific computing team wants to understand how a mathematical approximation changes as more terms are included.

We will use NumPy to create a sequence, calculate a trigonometric component, generate series terms, and observe convergence.

## Task 1 - Define the Angle

In [28]:
theta = 30

theta_radians = np.radians(theta)

print(f"Angle in Degrees: {theta}")
print(f"Angle in Radians: {theta_radians:.2f}")

Angle in Degrees: 30
Angle in Radians: 0.52


## Task 2 - Create the Sequence of Terms

In [37]:
k = np.arange(1, 100, 1)

print(f"First Value: {k[0]}")
print(f"Last Value: {k[-1]}")
print(f"Number of Values Generated: {len(k)}")

First Value: 1
Last Value: 99
Number of Values Generated: 99


## Explanation

The sequence starts at 1 and stops before 100.

Because the stop value in np.arange() is not included, the last value is 99.

Therefore, the sequence contains 99 values.

## Task 3 - Calculate the Trigonometric Component

In [39]:
sin_theta = np.sin(theta_radians)

print(f"sin(theta) = {sin_theta:.4f}")

sin(theta) = 0.5000


## Task 4 - Build the Series Terms

In [41]:
series_terms = sin_theta / k

print(f"Series Terms: {series_terms}")

Series Terms: [0.5        0.25       0.16666667 0.125      0.1        0.08333333
 0.07142857 0.0625     0.05555556 0.05       0.04545455 0.04166667
 0.03846154 0.03571429 0.03333333 0.03125    0.02941176 0.02777778
 0.02631579 0.025      0.02380952 0.02272727 0.02173913 0.02083333
 0.02       0.01923077 0.01851852 0.01785714 0.01724138 0.01666667
 0.01612903 0.015625   0.01515152 0.01470588 0.01428571 0.01388889
 0.01351351 0.01315789 0.01282051 0.0125     0.01219512 0.01190476
 0.01162791 0.01136364 0.01111111 0.01086957 0.0106383  0.01041667
 0.01020408 0.01       0.00980392 0.00961538 0.00943396 0.00925926
 0.00909091 0.00892857 0.00877193 0.00862069 0.00847458 0.00833333
 0.00819672 0.00806452 0.00793651 0.0078125  0.00769231 0.00757576
 0.00746269 0.00735294 0.00724638 0.00714286 0.00704225 0.00694444
 0.00684932 0.00675676 0.00666667 0.00657895 0.00649351 0.00641026
 0.00632911 0.00625    0.00617284 0.00609756 0.0060241  0.00595238
 0.00588235 0.00581395 0.00574713 0.00568182 0.0

## Explanation

NumPy performs element-wise division.

The single value sin(theta) is divided by every value in the k array.

Therefore, the operation produces one series term for each value of k.

## Task 5 - Calculate the Sum

In [43]:
series_sum = np.sum(series_terms)

print(f"Sum of the Generated Terms: {series_sum:.3f}")

Sum of the Generated Terms: 2.589


## Task 6 - Increase the Number of Terms

In [50]:
k_first = np.arange(1, 100, 1)
terms_first = sin_theta / k_first
print(f"Sum of the First Terms: {np.sum(terms_first)}")

k_second = np.arange(1, 1000, 1)
terms_second = sin_theta / k_second
print(f"Sum of the Second Terms: {np.sum(terms_second)}")

k_third = np.arange(1, 10000, 1)
terms_third = sin_theta / k_third
print(f"Sum of the Third Terms: {np.sum(terms_third)}")

Sum of the First Terms: 2.5886887588198104
Sum of the Second Terms: 3.7422354302751724
Sum of the Third Terms: 4.893753018022191


## Task 7 - Investigate Convergence

## Convergence Results

| Number of Terms | Series Result |
|---:|---:|
| 99 | 2.5886887588 |
| 999 | 3.7422354303 |
| 9,999 | 4.8937530180 |

As the number of terms increases, the series result also increases. The approximation changes because more terms are being added to the series.

## Task 8 - Explain Convergence

## Explanation of Convergence

### 1. What happened when fewer terms were used?

When fewer terms were used, the calculated series result was smaller because fewer terms were included in the calculation.

### 2. What happened when the number of terms increased?

When the number of terms increased from 99 to 999 and then to 9,999, the series result increased.

### 3. Why does increasing the number of terms change the approximation?

Increasing the number of terms adds more values to the calculation. Since the additional terms contribute to the total, the resulting approximation changes.

### 4. What does it mean when the result gets closer to a limiting value?

It means that the calculation is approaching a value that the series tends toward as more terms are included. This behavior is known as convergence.

# BONUS CHALLENGE
## Sports Performance Analysis

### Scenario

A football player records the number of successful passes made in five matches.

The goal is to use NumPy to analyze the player's passing performance and calculate the total, average, improvement, and percentage-related information.

### Objective

The objective is to use NumPy mathematical operations to analyze the player's performance across five matches.

In [53]:
matches = np.array([32, 38, 35, 42, 45])

print(f"Successful Passes: {matches}")

Successful Passes: [32 38 35 42 45]


## Bonus Calculation 1 - Total Successful Passes

In [55]:
total_passes = np.sum(matches)

print(f"Total Successful Passes: {total_passes}")

Total Successful Passes: 192


## Bonus Calculation 2 - Average Successful Passes

In [56]:
average_passes = np.mean(matches)

print(f"Average Successful Passes: {average_passes}")

Average Successful Passes: 38.4


## Bonus Calculation 3 - Difference from Average

In [58]:
pass_deviations = matches - average_passes

print(f"Deviations from Average: {pass_deviations}")

Deviations from Average: [-6.4 -0.4 -3.4  3.6  6.6]


## Bonus Calculation 4 - 10% Performance Increase

In [60]:
projected_passes = matches * 1.10

print(f"Projected Passes After 10% Increase: {projected_passes}")

Projected Passes After 10% Increase: [35.2 41.8 38.5 46.2 49.5]


## Bonus Calculation 5 - Difference Between Projected and Original

In [62]:
additional_passes = projected_passes - matches

print(f"Additional Projected Passes: {additional_passes}")

Additional Projected Passes: [3.2 3.8 3.5 4.2 4.5]


## Results

The player recorded 192 successful passes across five matches.

The average number of successful passes per match was 38.4.

The deviations show how each match compared with the average. The first match was 6.4 passes below the average, while the fifth match was 6.6 passes above the average.

If the player's successful passes increased by 10%, the projected performances would range from 35.2 to 49.5 successful passes per match.

## Interpretation

The results show that the player's passing performance varied across the five matches. The average of 38.4 provides a general measure of performance, while the deviations show how individual matches differed from that average.

The 10% projection also demonstrates how NumPy can be used to apply the same mathematical operation to an entire dataset.

## Conclusion

NumPy makes it possible to perform calculations on multiple numerical values efficiently. In this experiment, NumPy was used to calculate the total and average performance, measure deviations from the average, and apply a hypothetical 10% increase to the dataset.

# Overall Conclusion

These exercises demonstrated how NumPy can be used to solve numerical problems in different real-world contexts.

In the Business Analytics exercise, NumPy was used to calculate total sales, average sales, adjusted sales, and additional revenue.

In the Education Analytics exercise, NumPy was used to calculate the mean, deviations, squared deviations, and sum of squared deviations.

In the Engineering and Scientific Computing exercise, NumPy was used to convert angles, create numerical sequences, calculate sine, build series terms, and investigate convergence.

The Bonus Challenge demonstrated how the same NumPy concepts can be applied to a sports performance scenario.

The main lesson is that NumPy is not only about writing functions such as np.sum() or np.mean(). It is about using numerical data to perform calculations, inspect results, and understand what those results mean in a real-world situation.